dask/layers.py

```diff
@@ -48,11 +55,22 @@ class CreateArrayDeps(BlockwiseDep):
 
     def __init__(self, chunks: tuple):
         self.chunks = chunks
+        self.shape = tuple(sum(c) for c in chunks)
+        self.starts = [cached_cumsum(c, initial_zero=True) for c in chunks]
         self.numblocks = tuple(len(chunk) for chunk in chunks)
         self.produces_tasks = False
 
     def __getitem__(self, idx: tuple):
-        return tuple(chunk[i] for i, chunk in zip(idx, self.chunks))
+        chunk_shape = tuple(chunk[i] for i, chunk in zip(idx, self.chunks))
+        array_location = tuple(
+            (start[i], start[i + 1]) for i, start in zip(idx, self.starts)
+        )
+        return {
+            "shape": self.shape,
+            "num-chunks": self.numblocks,
+            "array-location": array_location,
+            "chunk-shape": chunk_shape,
+        }
```

Comment:
Going all the way back to discussion in https://github.com/dask/dask/issues/7513, I'd prefer a compositional/unix-like approach to these BlockwiseDeps. Rather than one `CreateArrayDeps` that returns a dict of possibly-useful things, I'd prefer a separate `ArrayLocationDep`, `ChunkShapeDep`, etc. Then when you're using blockwise, you can compose together whatever helpers you need for the appropriate function.

This would eliminate `_get_from_block_info`, `_create_uniform_array_from_block_info`, etc. You'd no longer need to pull the fields out of the dict, you'd just get them passed directly into the function. This even eliminates the need for `partial`s and wrapper functions. If you have a BlockwiseDep like [`Slices`](https://github.com/dask/dask/blob/a154dae7f9893965a95ecd3dde6ec0fa6868764c/dask/blockwise.py#L259-L276), you can just use the normal `getitem` as your function, the normal `arr` as argument 1 (a literal), and `Slices(chunks)` as argument 2—now you're just calling `getitem(arr, slices)`.

I might then vote to eliminate `BlockwiseCreateArray` entirely. It's a thin wrapper that may be more indirection than it's worth.

Once you get the core concept of blockwise—it's like NumPy broadcasting & vectorization, but applied to chunks of graphs instead of rows of arrays—I personally think it's easier to use the blockwise interface directly, rather than understanding how wrappers like `BlockwiseCreateArray` adjust that interface.

`BlockwiseCreateArray` is very limiting, in that you can't pass in any other blockwise arguments to it. If you want to create a blockwise array that depends on some other static argument (exactly what `inline_array=False` requires), you have to use plain `blockwise` directly. (Sure, you can extend `BlockwiseCreateArray` to take some `**extra_indices`, but what if you need to control where the `CreateArrayDeps` falls in the order of the arguments? Frankly, the only nice thing that `BlockwiseCreateArray` does for you is `out_ind = tuple(range(len(shape)))`. And if you're passing in your own `extra_indices`, you have to generate those `out_ind` yourself anyway, so you lose even that little convenience.)

Once there are lots of little `BlockwiseDep` tools you can compose together (kinda like `itertools` for blockwise), it'll make more sense for each instance of array creation to just call `blockwise` directly, and pass in the appropriate helpers for that situation, than to have a monolithic `BlockwiseCreateArray` interface that tries to satisfy the needs of everything.

Reply:
I took a crack at splitting the `CreateArrayDeps` into separate helpers, and ultimately I agree with you, it's nicer this way. It's also nice to remove `BlockwiseCreateArray`.

The downside, here, is that all of this is almost entirely undocumented! I actually wasn't aware that `dask.blockwise.blockwise()` could take `BlockwiseDep` objects and do the right thing. But, as far as I can tell, `dask.array.blockwise.blockwise()` *cannot* take `BlockwiseDep` objects! And neither of them really describe what are valid inputs. It doesn't really make me feel great about the ability to point other people at this and say "make more high-level-graphs!".

Reply:
I'm also liking this a lot more now! I much prefer the `ArraySliceDep` pattern, and it's nice how easy the inline vs non-inline case became.

> The downside, here, is that all of this is almost entirely undocumented!...It doesn't really make me feel great about the ability to point other people at this and say "make more high-level-graphs!".

Agreed! To be fair though, almost nothing about HLGs or Blockwise is documented (https://github.com/dask/dask/issues/7709). I feel like other people already can't make high-level graphs, regardless of how we handle `BlockwiseDeps` here. As we've discussed, the patterns for HLGs both aren't very well established, and aren't that great yet. On the bright side, the lack of documentation or significant external use gives us a lot of leeway to iterate on those patterns right now and adjust towards better ones, so that when we do document and clean them up, we're presenting something that's easy to use and understand.

> But, as far as I can tell, `dask.array.blockwise.blockwise()` _cannot_ take `BlockwiseDep` objects!

Correct, and it's annoying. I made `DataFrame.map_partitions` support them in https://github.com/dask/dask/pull/8310, and there was work to make `da.blockwise` support them in https://github.com/dask/dask/pull/7686, so I think that's definitely the way things are headed. We should make `da.blockwise` support them too.